### ISSUE TITLE
Opinion on adding embed.text? Any way to get "current card" info?

### ISSUE THREAD

**Issue by spdustin** [andymeneely/squib#82]
First off, awesome project. Just saw it on HN this morning, and already started making a custom [Fluxx](http://www.looneylabs.com/games/fluxx) deck for my son's grand idea: A Harry Potter (inspired, for licensing purposes) Fluxx game. (The screenshot is very much a work in progress - default flavor text is missing from most of these cards, but I wanted to see how far I could get in a couple of hours with `squib` before committing to being "data-complete" :stuck_out_tongue_winking_eye:)

![sheet00_resized](https://cloud.githubusercontent.com/assets/1342542/8417781/4cacf2a4-1e74-11e5-8a4e-174b125fa2ad.png)

I'll plan to post that to a new repo here on Github when it's done, but here's the query: `TextEmbed` handles PNGs or SVGs - what about text replacements? As it stands, there doesn't seem to be a "*squib*" way to alter the text of the card in a `deck` on a per-card basis. I know that most of the pre-processing you'd want to do you could just put into the source data - things like formatting options. That can make one repeat oneself, though! And one additional thing I wanted to do is automatically bold any occurrence of a card title in the card text - in my screenshot, "**The Seer**" includes in its text a reference to "**Hourglass**", which was **bolded** automatically using the code I've enclosed below.

What I'm doing instead is using `collect` to "pre-process" the deck object. Do you have an opinion on pre-processing versus extending `TextEmbed` to support in-place text replacements? Or maybe refactoring in-place replacements altogether to a more "ruby-esque" collect-and-replace pre-processing of the `deck` object?

```ruby
words_to_bold = data['Title'] + [
  'keeper[s]?','creeper[s]?','action','new rule',
  'goal','surprise','draw pile','discard pile',
  'your hand','magic user'
]
data['Text'].collect! { |thetext|
  unless !thetext
    thetext.gsub(/(#{(words_to_bold).join '|'})/i,'<b>\1</b>')
  end
}
```

If there's a way to reference the current item --by its index maybe? -- **squib** would then support dynamically referencing a layout section using metadata from the current card. Right now, I have six different `rect range:` statements using the sample you gave for creating an index of cards based on a field - those ranges are written out for each card type so they can reference different sections of the layout file, but I would REALLY love to do something like `layout: "sidebar_#{data['Type']}"` to choose the layout section used for a given type of card. I did parse thru the source to see if I could quickly grok how to do that, but it wasn't immediately evident. I'm no Ruby pro, so I could be missing something entirely.

Any ideas on that last paragraph? Or opinions on processing the text of a card in advance versus in-line?

**Comment by spdustin** [comment]
On that last point (referencing current card to dynamically choose a layout) here's how I'm currently handling things:
```ruby
type = {}; data['Type'].each_with_index{ |t,i| (type[t] ||= []) << i}
rect range: type['Creeper']  , layout: 'sidebar_creeper'
rect range: type['Action']   , layout: 'sidebar_action'
rect range: type['Keeper']   , layout: 'sidebar_keeper'
rect range: type['New Rule'] , layout: 'sidebar_newrule'
rect range: type['Goal']     , layout: 'sidebar_goal'
rect range: type['Surprise'] , layout: 'sidebar_surprise'
text range: type['Creeper']  , layout: 'sidebar_text_creeper'  , str: data['Title']
text range: type['Action']   , layout: 'sidebar_text_action'   , str: data['Title']
text range: type['Keeper']   , layout: 'sidebar_text_keeper'   , str: data['Title']
text range: type['New Rule'] , layout: 'sidebar_text_newrule'  , str: data['Title']
text range: type['Goal']     , layout: 'sidebar_text_goal'     , str: data['Title']
text range: type['Surprise'] , layout: 'sidebar_text_surprise' , str: data['Title']
```
That's the part I'd LOVE to refactor to:
```ruby
type = {}; data['Type'].each_with_index{ |t,i| (type[t] ||= []) << i}
rect layout: "sidebar_#{data['Type']}"
text layout: "sidebar_#{data['Type']}", str: data['Title']
```

**Comment by andymeneely** [comment]
Glad you like it! Those already look like real Fluxx cards. If you post it on GitHub, I'll link to your final product in a showcase section on the website (if you need an incentive...)

What I would do for processing strings would look a lot like your approach. With some refactoring (such as declaring a helper method), it could look pretty clean. 

I hadn't thought about providing that with `TextEmbed` though. Hm. I'll have to think on that. Something like this? 
```ruby
text(str: data['Rule']) do |embed|
  embed.embolden words: %w(creeper action)
end
```

Another place would be in reading from xlsx or csv. I've run into issues where Excel (via the Roo parser we use) provides only floating-point numbers and it would be nice to cleanly convert it with `to_i` when we parse. 

Regarding the `rect range` ones. `layout:` does take arrays that correspond to the card. So you could do:

```ruby
rect layout: data['Type'].map { |t| "sidebar_#{t}" }
```

**Comment by spdustin** [comment]
Re: `rect range`, that's excellent. I made one small change for my build (because the "Type" is a Title Case string with spaces, and the matching layout is a lowercased slug with no spaces). I was able to reduce those thirteen lines in my previous comment to this:

```ruby
type = {}; data['Type'].each_with_index{ |t,i| (type[t] ||= []) << i}
rect layout: data['Type'].map { |t| "sidebar_#{t.downcase.sub(/\s/,'')}" }
text str: data['Type'].map { |t| t.upcase }, layout: 'card_title'
```

Works really well - thanks a ton!

I tried the same approach on one more repeating section of the code, and got an error, but I'll post another issue for that.

I'll noodle on the pre-processing approaches - let's keep the conversation going on it, I'd love to help out. And, yeah, it'd be great to get the project listed once we're done. With **squib**, it should go pretty quickly. :smile:
